server/cache/dirtools/dirtools.go

```diff
@@ -980,37 +1058,78 @@ func (ff *BatchFileFetcher) bytestreamReadFiles(ctx context.Context, instanceNam
 		ff.stats.FileDownloadCount += 1
 		ff.statsMu.Unlock()
 
-		if err := f.Close(); err != nil {
-			return nil, err
+		if err := w.Commit(); err != nil {
+			return nil, status.WrapError(err, "could not commit byte stream input writer")
 		}
-		fileCache := ff.env.GetFileCache()
-		if fileCache != nil {
-			if err := fileCache.AddFile(ff.ctx, fp0.FileNode, fp0.FullPath); err != nil {
-				log.Warningf("Error adding file to filecache: %s", err)
+
+		if err := w.Close(); err != nil {
+			return nil, status.WrapError(err, "could not close byte stream input writer")
+		}
+
+		if !ff.onlyDownloadToFileCache {
```

Comment:
I find the `if !ff.onlyDownloadToFileCache {` logic sprinkled throughout kind of difficult to follow, but I'm not sure I have any suggestions about how to improve it due to how this code is structured. I guess if you had methods this stuff your main algorithm flow could just call those methods and you could short-circuit from within them if you were in filecacheOnly mode.